onnx/onnx-ml.proto

```diff
@@ -188,6 +194,115 @@ message NodeProto {
+
+  // The attributes of the function.
+  repeated string attribute= 6;
+
+  // The nodes in the function.
+  repeated NodeProto node = 7;
+  // A human-readable documentation for this function. Markdown is allowed.
+  optional string doc_string = 8;
+}
+
+// Training information
+// TrainingInfoProto contains the minimized loss function (e.g., mean
+// squared error), the used optimization algorithm (a mapping from gradient
+// and the current tensors to their new values; that is, one training iteration),
+// and the binding between the optimized tensors and their gradient tensors'
+// names (we need to give a name (e.g., "gradient_of_X") to the gradient of
+// tensor "X", so that gradient tensors can be referenced in the training 
+// algorithm).
+//
+// This design is mainly designed for popular stochastic gradient-based methods.
+// Conceptually, a TrainingInfoProto defines a "partial computation graph"
+// which will be merged into the inference graph to form a full training
+// computation graph.
+message TrainingInfoProto {
+  // Training-specific initializers, for example, accumulated squared gradient
+  // in ADAGRAD optimizer, momentum, or number of executed training iterations.
+  // The values can only be inputs of "loss" and "optimizer" defined below.
+  repeated TensorProto additional_initializer = 1;
+
+  // Training-specific inputs, for example, labels in classification problems.
+  // The values can only be inputs of "loss" and "optimizer" defined below.
+  // The full input list of the training computation graph is the concatenation
+  // of "ModelProto.graph.input" and
+  // "ModelProto.training_info.additional_input".
+  repeated ValueInfoProto additional_input = 2;
+
+  // Training-specific outputs, for example, new tensor values computed by the
+  // specified training algorithm. The full output list of the training
+  // computation graph is the concatenation of "ModelProto.graph.output" and
+  // "ModelProto.training_info.additional_output".
+  repeated ValueInfoProto additional_output = 3;
+  
+  // Loss node which consumes tensors in "ModelProto.graph", "additional_inputs"
+  // , and "additional_initializer". Loss nodes produces a scalar loss value.
+  // In training phase, the optimization algorithm would be launch to minimizes
+  // loss node's output value. "loss" can be an ONNX primitive operator or a
+  // function defined in "ModelProto.function."
+  required NodeProto loss = 4;
+  
+  // Optimization algorithm used to minimize the loss function. It can be an
+  // ONNX primitive operator or a function defined in "ModelProto.function."
+  // This node can reference tensors in the "ModelProto.graph",
+  // "additional_inputs," "additional_initializer," and gradient tensor names
+  // defined below.
+  // To connect the optimizer's parameter, optimized tensors, optimized tensors'
+  // gradient, and optimizer correctly without breaking
+  // single-static-assignment, "optimizer" has some restruction on its input
```

Comment:
I don't think we need the restrictions below now, do we? If we have pre-defined ONNX ops (for the optimizer), the op spec will indicate the required operand-ordering. We can recommend the following ordering as a preferred practice, but it is  not clear to me why this is a compulsory requirement. (Also: spelling correction "restruction" => "restriction" … but I don't think we need restrictions.)

Reply:
They are for customized training algorithms. As you mentioned, pre-defined ONNX optimizers have some kind of special schemas for correcting wiring tensors. For the same reason, customized optimizers (defined as FunctionProto in `ModelProto.function`) should use the same schema.

Reply:
Yes, but the runtime doesn't care really since we already capture the relationship between optimized-tensors and their-gradients separately, and the optimizer NodeProto's input list captures the relationship between the tensors/gradients computed in the preceding steps and the inputs required by the optimizer custom-function. So, I don't understand what extra-restrictions we need to talk about. (There is already an implicit requirement that the inputs listed in the optimizer NodeProto's input should be consistent with the input-ordering in the optimizer custom-function/onnx-op.)

Reply:
In other words, there is no requirement that different customized optimization functions should all follow the same pattern/style for ordering inputs/outputs. However, if they all follows the same pattern/style, it may end up being helpful in the future (or to users). So, I can understand the value in a "recommended ordering" (but it doesn't seem like it needs to be a "mandated ordering").

Reply:
Let's consider an example. Can user specify an optimization function which outputs `[Gradient_X1, new_X1, new_X2, Gradient_X2]`? How would runtime know "new_X1" is the new value of input tensor "X"? In addition, there can be other FunctionProto optimizers which produce `[A, B, C, D]`, `[C, A, B, D]`, and so on. Can runtime bind `[A, B, C, D]` correctly to updated tensors and updated states without those restrictions?

Reply:
new_X1 would be in the additional_input list (of TrainingInfoProto), right? It will also appear in the optimizer NodeProto's input list. (The relationship between items in these lists are simply by name.) A corresponding formal_parameter (call it formal_new_X1) will appear in the customized FunctionProto. The correspondence between items in the Optimizer NodeProto input list and the customized FunctionProto input list is by position. So, what else does the runtime need to know?

Reply:
Sorry, new_X1 would probably in the inference graph's input list. Other things (like state_X1) may be in the additional_input list. It doesn't matter. But they are referred to by name in the optimizer NodeProto's input list.

Reply:
Sorry again (several typos!): I am talking about inputs, but the same thing applies for outputs. new_X1 in optimizer NodeProto's output list is associated with the additional_outputs of training ny name.

Reply:
Spoke with Wei-sheng offline. Now I understand his goals. It sounds like this is more a schema restriction for the training-graph (rather than the optimizer function). Can we specify the restriction that the additional_training_outputs must correspond to the concat(graph initializers, training graph additional initializers) one-to-one?

Reply:
Otherwise, we may need to add the mapping as another field.

Reply:
One-to-one looks too much because not every tensor is updated in the training phase. Let's add another string-to-string entry list.